_posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd

```diff
@@ -0,0 +1,326 @@
+bibliography: bibliography.bib
+date: 05-29-2020
+categories:
+  - R
+  - Image Recognition & Image Processing
+  - TensorFlow/Keras
+  - Probabilistic ML/DL
+  - Unsupervised Learning
+output:
+  distill::distill_article:
+    self_contained: false
+preview: images/thumb.png
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+We've seen quite a few examples of unsupervised learning (or self-supervised learning, to choose the more correct but less
+popular term) on this blog.
+
+Often, these involved *Variational Autoencoders (VAEs),* whose appeal lies in them allowing to model a *latent space* of
+underlying, independent (preferably) factors that determine the visible features. A possible downside can be the inferior
+quality of generated samples. Another popular approach are GANs (Generative Adversarial Networks). Conceptually, these are
+highly attractive due to their game-theoretic framing. However, they can be difficult to train. *PixelCNN* variants, on the
+other hand -- we'll subsume them all here under PixelCNN -- are generally known for their good results. They seem to involve
+some more alchemy[^1] though. Under those circumstances, what could be more welcome than an easy way of experimenting with
+them? Through TensorFlow Probability (TFP) and its R wrapper, [tfprobability](https://github.io/tfprobability), we now have
+such a way.
+
+[^1]: Alluding to Ali Rahimi's (in-)famous "deep learning is alchemy" [talk](https://www.youtube.com/watch?v=Qi1Yry33TQE) at
+    NeurIPS 2017. I would suspect that to some degree, that statement resonates with many DL practitioners -- although one
+    need not agree that more mathematical rigor is the solution.
+
+This post first gives an introduction to PixelCNN, concentrating on high-level concepts (leaving the details to the curious,
+to look them up in the respective papers). We'll then show an example of using `tfprobability` to experiment with the TFP
+implementation.
+
+## PixelCNN principles
+
+### Autoregressivity, or: We need (some) order
+
+The basic idea in PixelCNN is autoregressivity. Each pixel is modeled as depending on all prior pixels. Formally:
+
+$$p(\mathbf{x}) = \prod_{i}p(x_i|x_0, x_1, ..., x_{i-1})$$
+
+Now wait a second - what even *are* prior pixels? Last I saw one images were two-dimensional. So this means we have to impose
+an *order* on the pixels. Commonly this will be *raster scan* order: row after row, from left to right. But when dealing with
+color images, there's something else: At each position, we actually have *three* intensity values, one for each of red, green,
+and blue. The original PixelCNN paper[@OordKK16] carried through autoregressivity here as well, with a pixel's intensity for
+red depending on just prior pixels, those for green depending on these same prior pixels but additionally, the current value
+for red, and those for blue depending on the prior pixels as well as the current values for red and green.
```

Comment:
Ah, so the "addition" is of another sort of "prior" value? I think you might be explaining this as well as is humanly possible, I just know very very little about this, and autoregressivity is hard to describe! 😆

Reply:
don't know - I think it's consistent in the context...